table_schema_cache.go

```diff
@@ -189,10 +190,7 @@ func LoadTables(db *sql.DB, tableFilter TableFilter, columnCompressionConfig Col
 			tableLog := dbLog.WithField("table", tableName)
 			tableLog.Debug("caching table schema")
 
-			// Sanity check
-			if len(tableSchema.PKColumns) != 1 {
-				err = fmt.Errorf("table %s has %d primary key columns and this is not supported", tableName, len(tableSchema.PKColumns))
-				logger.WithError(err).Error("invalid table")
+			if err := tableSchema.setPaginationKeyColumn(dbname, tableName, paginateByColumnConfig); err != nil {
```

Comment:
A little philosophical but does it make sense to figure out the column before even constructing the new `TableSchema`s?

i.e. on L176 we have `tableSchemas = append(tableSchemas, &TableSchema{...})`. We could infer the pagination column before even calling `append`

Reply:
There are multiple perspectives to this:
1. Should we adopt an optimistic or pessimistic outlook on user supplied configuration? For the optimistic outlook, you make the assumption that the user supplied sane configuration unless proven otherwise. For the pessimistic outlook you make the inverse assumption. 

If you adopt the optimistic outlook, then you basically want to make the usage as permissible as possible and thus you accommodate for the scenario where the user supplied incorrect configuration but which never harms the system as it is not being applied. e.g.:
The line: `tableSchemas, err = tableFilter.ApplicableTables(tableSchemas)` filters out any tables that are ignored. In the case that the user supplied incorrect configuration for a table that he has blacklisted, then because it gets filtered out, the subsequent loop will never error out for `tableSchema.setPaginationKeyColumn` as the loop simply skips that table.

In the pessimistic outlook, you would want to run the sanity checks for all tables, whether or not blacklisted. 

It is the same 'being-optimistic-vs-pessimistic-about-user-input' debate you would have with regards to the design of a programming language whether you want it to be a dynamically typed language or a statically typed one. 
 
Which outlook do you feel is best for this project? From my side, I mainly placed it here as the previous sanity check hinted to me that the project adopted an optimistic outlook on user configuration supplied:

```
if len(tableSchema.PKColumns) != 1 {	
				err = fmt.Errorf("table %s has %d primary key columns and this is not supported", tableName, len(tableSchema.PKColumns))	
				logger.WithError(err).Error("invalid table")	
				return tableSchemaCache, err
```

Likewise the function introduced by me `setPaginationKeyColumn` adopts the `optimistic` outlook by it only running on the tables that are effectively whitelisted

In a `pessimistic` outlook I would promote the same to the other loop so it will always error out despite the table being blacklisted.

2. How explicit do we want to be with our code base to facilitate readability at first glance?
We declared a type: 
```
type TableSchema struct {
	*schema.Table

	CompressedColumnsForVerification map[string]string   // Map of column name => compression type
	IgnoredColumnsForVerification    map[string]struct{} // Set of column name
	PaginationKeyColumn *schema.TableColumn

	rowMd5Query string
}
```

If we want to be explicit, then the following code:
```			
tableSchemas = append(tableSchemas, &TableSchema{
				Table: tableSchema,
				CompressedColumnsForVerification: columnCompressionConfig.CompressedColumnsFor(dbname, table),
				IgnoredColumnsForVerification: columnIgnoreConfig.IgnoredColumnsFor(dbname, table),
			})
```
can be a source of confusion where a developer will have to look through the code base to see where the missing fields are being set. It would be more clearer to have `PaginationKeyColumn` to be explicitly set as:
```			tableSchemas = append(tableSchemas, &TableSchema{
				Table: tableSchema,
				CompressedColumnsForVerification: columnCompressionConfig.CompressedColumnsFor(dbname, table),
				IgnoredColumnsForVerification: columnIgnoreConfig.IgnoredColumnsFor(dbname, table),
				PaginationKeyColumn: paginationKeyColumn,
			})
```
If we adopt the pessimistic outlook, then we could state:
```
if paginationKeyColumn, err = GetPaginationKeyColumn(...); err != nil {
...
}
```
If we adopt the optimistic outlook then we could state:
```			
tableSchemas = append(tableSchemas, &TableSchema{
				Table: tableSchema,
				CompressedColumnsForVerification: columnCompressionConfig.CompressedColumnsFor(dbname, table),
				IgnoredColumnsForVerification: columnIgnoreConfig.IgnoredColumnsFor(dbname, table),
				PaginationKeyColumn: nil // will be set shortly,
			})
```
Do you feel that being explicit facilitates better code maintainability? 
What about the key: `rowMd5Query`? Should we also set it explicitly:
```			
tableSchemas = append(tableSchemas, &TableSchema{
				Table: tableSchema,
				CompressedColumnsForVerification: columnCompressionConfig.CompressedColumnsFor(dbname, table),
				IgnoredColumnsForVerification: columnIgnoreConfig.IgnoredColumnsFor(dbname, table),
				PaginationKeyColumn: nil // will be set shortly,
                                   rowMd5Query: "" // will be set later when verification being done,
			})
```

Lastly, I made `PaginationKeyColumn` public. Would it make sense for it to be private instead?

Reply:
Reading the code again, the whole `LoadTables` function needs a cleanup and this PR is not making it any worse than before. I don't really wanna block the PR on this discussion so it's fine as is, though improving the readability of the `setPaginationKeyColumn` function still seems like a good idea.